crates/wit-parser/src/ast.rs

```diff
@@ -82,60 +82,162 @@ impl<'a> DeclList<'a> {
         }
     }
 
-    fn parse_implicit_package_items(
+    fn parse_main_package_items(
         tokens: &mut Tokenizer<'a>,
         mut docs: Docs<'a>,
-    ) -> Result<DeclList<'a>> {
+        mut nested: Vec<NestedPackage<'a>>,
+        main_pkg_decl: Option<&PackageName>,
+    ) -> Result<(DeclList<'a>, Vec<NestedPackage<'a>>)> {
         let mut items = Vec::new();
         while tokens.clone().next()?.is_some() {
-            items.push(AstItem::parse(tokens, docs)?);
-            docs = parse_docs(tokens)?;
+            if tokens.eat(Token::Package)? {
+                let package_id = PackageName::parse(tokens, &mut docs)?;
+                if let Some(main) = main_pkg_decl {
+                    if main.package_name() == package_id.package_name() {
+                        bail!(
+                            "Nested package name already used by main package declaration: {}",
+                            main.package_name()
+                        );
+                    }
+                }
```

Comment:
Could this error get deferred to somewhere in `ast/resolve.rs` perhaps? Ideally the errors in this file are just parse errors rather than semantic errors.